### PR TITLE
add permissions service for overview router. add guard to restrict ac…

### PR DESCRIPTION
**Pull request by whitfield-mj**
…cess based on service

### CODE REVIEW

File: server/controllers/overviewController.test.ts
Comment (brightonsbox) [inline server/controllers/overviewController.test.ts:157]: Was in two minds whether we could just use `res` instead of explicitly having the `resNotInCaseload`... since it's now relying on the permissions, it's not necessary is it and bloats the test... meh...

File: server/controllers/overviewController.ts
Comment (brightonsbox) [inline server/controllers/overviewController.ts:96]: This has changed from just requiring `prisonerInCaseLoad` to `isPrisonerInCaseLoad || userHasRoles([Role.PomUser, Role.ReceptionUser], userRoles)` 

https://github.com/ministryofjustice/hmpps-prisoner-profile/pull/658/files#diff-436cd1a03284961f6db546c6d570d7f2d99d017aacdcf407b84fb925c79cfa47R60

Is this right?

File: server/services/permissionsService.test.ts
Comment (brightonsbox) [inline server/services/permissionsService.test.ts:310]: I think I perhaps preferred the original parameterised test style for this, think I find it easier to read: 

https://github.com/ministryofjustice/hmpps-prisoner-profile/pull/658/files#diff-f2abab7599c1a00864d045e892ea14706179469fe03671a1ef4e8545fcc891faL39-L50

But that's just personal preference, so happy to let it slide, it looks like all the cases are covered

File: server/services/permissionsService.test.ts
Comment (brightonsbox) [inline server/services/permissionsService.test.ts:335]: Could add `... regardless of case load` in the description perhaps?

File: server/services/permissionsService.test.ts
Comment (brightonsbox) [inline server/services/permissionsService.test.ts:339]: Or I wonder if the intent would be clearer if it was `caseLoadId: 'SOMETHING_ELSE'` since at first glance we can't see that it's actually different from the prisoner mock data 🤷

File: server/services/permissionsService.ts
Comment (brightonsbox) [reply server/services/permissionsService.ts:82]: (same with the other hardcoded `view: false` instances)

File: server/services/permissionsService.test.ts
Comment (brightonsbox) [inline server/services/permissionsService.test.ts:365]: I think we're missing a check of `should return edit false if prisoner is out` (`and user does not have inactive bookings role`)

Comment (brightonsbox) [reply server/services/permissionsService.test.ts:365]: Same again for transfer?

I know they're effectively just like any other case load, so if you don't have them you haven't got access, but it might be worth explicitly calling out in the tests?

File: server/services/permissionsService.ts
Comment (whitfield-mj) [reply server/services/permissionsService.ts:82]: I had assumed that they would all require a view at some point further into this process and was trying to avoid a type where everything is optional as it is pointless. Maybe they won't though so can take out for now

File: server/controllers/overviewController.test.ts
Comment (whitfield-mj) [reply server/controllers/overviewController.test.ts:157]: Yeah probably right. The reason I left it was to give an idea to the reader why the permission would be false... but I guess that's not really the point of a unit test

File: server/controllers/overviewController.ts
Comment (whitfield-mj) [reply server/controllers/overviewController.ts:96]: No, looks like an oversight

File: server/services/permissionsService.test.ts
Comment (whitfield-mj) [reply server/services/permissionsService.test.ts:335]: ok

File: server/services/permissionsService.test.ts
Comment (whitfield-mj) [reply server/services/permissionsService.test.ts:365]: If the prisoner is out and the user doesn't have that role they won't get any permissions, it will bail out at the access section  tested [here](https://github.com/ministryofjustice/hmpps-prisoner-profile/pull/658/files#diff-95688007aec6993ad05ed72087f054153d07188c542f7d7af6f5557d1a305902R80)

File: server/services/permissionsService.test.ts
Comment (whitfield-mj) [reply server/services/permissionsService.test.ts:492]: Same as above

File: server/services/permissionsService.test.ts
Comment (brightonsbox) [inline server/services/permissionsService.test.ts:657]: TODO raise with Joe Cartwright and team?

File: server/controllers/utils/overviewController/buildOverviewActions.test.ts
Comment (brightonsbox) [inline server/controllers/utils/overviewController/buildOverviewActions.test.ts:182]: Since there are so many of these, it might be nice to have a little helper function:

`expectActionVisible(url, visible)`
?

File: server/controllers/utils/overviewController/buildOverviewActions.ts
Comment (brightonsbox) [inline server/controllers/utils/overviewController/buildOverviewActions.ts:64]: For JB benefit: I've got to here in the review 😅

File: server/controllers/utils/overviewController/buildOverviewActions.test.ts
Comment (whitfield-mj) [reply server/controllers/utils/overviewController/buildOverviewActions.test.ts:182]: Jon now you want me to fix everyone else's work!

File: server/services/permissionsService.test.ts
Comment (brightonsbox) [inline server/services/permissionsService.test.ts:132]: Do we want to test that it will return false otherwise? I guess I could go and just change the code to:

```
courtCases: {
  view: true,
  edit: true,
}
```

and these tests would still pass

File: server/services/permissionsService.test.ts
Comment (brightonsbox) [inline server/services/permissionsService.test.ts:205]: Test the view false case

File: server/services/permissionsService.test.ts
Comment (brightonsbox) [inline server/services/permissionsService.test.ts:237]: Test the view false case

File: server/services/permissionsService.test.ts
Comment (brightonsbox) [inline server/services/permissionsService.test.ts:258]: test view false case

File: server/services/permissionsService.test.ts
Comment (whitfield-mj) [reply server/services/permissionsService.test.ts:657]: Have raised with Joe Cartwright. As it is the same as what is currently there I don't think it is a reason to hold up this particular PR, but when I get a response I will update accordingly

File: server/services/permissionsService.test.ts
Comment (brightonsbox) [reply server/services/permissionsService.test.ts:365]: Agreed, but I shouldn't this permissionsService be tested independently of the permissionsGuard? Just thinking since in theory this could be ported over to a separate service sometime in the far future and it'd be good to have a complete set of tests

File: server/services/permissionsService.test.ts
Comment (brightonsbox) [inline server/services/permissionsService.test.ts:752]: Think we're missing a test case `should return view true if user has %s role and prisoner is %s`

with OUT and TRN as the final options. since it looks like if they've got the right role and the prisoner is out or transferring they can still view

File: server/services/permissionsService.test.ts
Comment (whitfield-mj) [reply server/services/permissionsService.test.ts:365]: Ok so that means I need to extract all of this and call separately

File: server/services/permissionsService.test.ts
Comment (brightonsbox) [inline server/services/permissionsService.test.ts:784]: Test the happy path?

**Review by brightonsbox — DISMISSED** [review]
This is excellent - as discussed approving so that you can follow up on the comments in a separate PR

**Review by brightonsbox — APPROVED** [review]
LGTM!!

**Pull request merged**
